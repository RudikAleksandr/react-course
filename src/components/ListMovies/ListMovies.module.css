
.listMoviesContainer .countMovies {
  color: white;
  font-size: 1.1rem;
  font-weight: bold;
  margin-right: 7px;
}

.listMoviesContainer .moviesFoundLabel {
  color: white;
  font-size: 1.1rem;
}

.listMovies {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(260px, 1fr));
  gap: 40px 75px;
}
