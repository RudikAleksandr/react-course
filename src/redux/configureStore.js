import { configureStore as configureStoreToolkit } from '@reduxjs/toolkit';
import thunkMiddleware from 'redux-thunk';
import moviesReducer from './moviesSlice';

const configureStore = (preloadedState) => (
  configureStoreToolkit({
    preloadedState,
    reducer: {
      movies: moviesReducer,
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
  })
);

export default configureStore;
