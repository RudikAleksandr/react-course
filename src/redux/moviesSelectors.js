
const selectListMovies = (state) => state.movies.moviesOrder.map(id => state.movies.movies[id]);

const selectSearch = (state) => state.movies.search;

const selectActiveMovieGenre = (state) => state.movies.activeMovieGenre;

const selectActiveSortType = (state) => state.movies.activeSortType;

export {
  selectListMovies,
  selectSearch,
  selectActiveMovieGenre,
  selectActiveSortType
};