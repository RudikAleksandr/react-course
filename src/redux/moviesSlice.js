import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as moviesAPI from '../api/moviesAPI';
import { groupByKey } from '../utils/grouping';

export const initialState = {
  movies: {},
  moviesOrder: [],
  search: '',
  activeMovieGenre: '',
  activeSortType: 'release_date'
};

export const getListMovies = createAsyncThunk(
  'movies/getListMovies',
  moviesAPI.getListMovies
);

export const getMovie = createAsyncThunk(
  'movies/getMovie',
  moviesAPI.getMovie
);

export const updateMovie = createAsyncThunk(
  'movies/updateMovie',
  moviesAPI.updateMovie
);

export const createMovie = createAsyncThunk(
  'movies/createMovie',
  moviesAPI.createMovie
);

export const deleteMovie = createAsyncThunk(
  'movies/deleteMovie',
  moviesAPI.deleteMovie
);

const moviesSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setSearch: (state, { payload }) => {
      state.search = payload;
    },
    setActiveMovieGenre: (state, { payload }) => {
      state.activeMovieGenre = payload;
    },
    setActiveSortType: (state, { payload }) => {
      state.activeSortType = payload;
    }
  },
  extraReducers: {
    [getListMovies.fulfilled]: (state, { payload }) => {
      state.movies = groupByKey('id', payload.data);
      state.moviesOrder = payload.data.map((movie) => movie.id);
    },
    [updateMovie.fulfilled]: (state, { payload }) => {
      state.movies[payload.id] = payload;
    },
  },
});

export const {
  setSearch,
  setActiveMovieGenre,
  setActiveSortType
} = moviesSlice.actions;

export default moviesSlice.reducer;