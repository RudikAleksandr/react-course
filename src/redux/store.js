import { configureStore } from '@reduxjs/toolkit';
import moviesReducer from './moviesSlice';
import thunkMiddleware from 'redux-thunk'

const store = configureStore({
  reducer: {
    movies: moviesReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
});

export default store;