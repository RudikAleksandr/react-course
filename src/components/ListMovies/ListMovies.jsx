import React, {
  useCallback, useEffect, useMemo, useState,
} from 'react';
import { Row, Col } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import MovieCard from '../MovieCard';
import cn from './ListMovies.module.css';
import MovieFormModal from '../MovieFormModal';
import MovieDeleteModal from '../MovieDeleteModal';
import { selectListMovies } from '../../redux/moviesSelectors';
import { deleteMovie, getListMovies } from '../../redux/moviesSlice';

const ListMovies = ({ search, genre, sortBy }) => {
  const [selectedOption, setSelectedOption] = useState({});
  const listMovies = useSelector(selectListMovies);
  const params = useMemo(() => ({
    search,
    sortBy,
    sortOrder: 'desc',
    searchBy: 'title',
    filter: genre,
  }), [search, genre, sortBy]);
  const dispatch = useDispatch();

  useEffect(() => {
    if (params.search || params.filter) {
      params.filter = params.filter === 'all' ? '' : params.filter;
      dispatch(getListMovies(params));
    }
  }, [dispatch, params]);

  // PATTERN: memoization of callbacks
  const handleSelectOption = useCallback((movie) => (type) => {
    setSelectedOption({ movie, type });
  }, []);

  const handleHideModal = useCallback(() => {
    setSelectedOption({});
  }, []);

  const handleConfirmDelete = useCallback(async () => {
    setSelectedOption({});
    await dispatch(deleteMovie(selectedOption.movie.id));
    dispatch(getListMovies(params));
  }, [dispatch, selectedOption, params]);

  return (
    <Row className={`flex-column mb-4 ${cn.listMoviesContainer}`}>
      <Col className="pl-0 mt-4 mb-4">
        <span className={cn.countMovies}>{listMovies.length}</span>
        <span className={cn.moviesFoundLabel}>movies found</span>
      </Col>
      <Col>
        <div className={listMovies.length ? cn.listMovies : cn.noMovieFound}>
          {/* PATTERN: Fragments */}
          {!!listMovies.length && listMovies.map((movie) => (
            <MovieCard
              onSelectOption={handleSelectOption(movie)}
              key={movie.id}
              movie={movie}
            />
          ))}
          {!listMovies.length && (
            <h2 className={cn.noMovieFoundLabel}>No Movie Found</h2>
          )}
        </div>
      </Col>
      {selectedOption.type === 'EDIT'
        && (
          <MovieFormModal
            movie={selectedOption.movie}
            onHide={handleHideModal}
          />
        )}
      {selectedOption.type === 'DELETE'
        && (
          <MovieDeleteModal
            onHide={handleHideModal}
            onConfirm={handleConfirmDelete}
          />
        )}
    </Row>
  );
};

export default ListMovies;
