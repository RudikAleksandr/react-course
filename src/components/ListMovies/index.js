import ListMovies from './ListMovies.jsx';

export default ListMovies;
