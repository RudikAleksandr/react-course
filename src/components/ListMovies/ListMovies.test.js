
import { render } from '@testing-library/react';
import ListMovies from './ListMovies';
import configureStore from 'redux-mock-store'
import { MemoryRouter, Route } from 'react-router';
import { getListMovies } from '../../redux/moviesSlice';
import * as reactRedux from 'react-redux';
import * as React from 'react';

jest.mock('../../redux/moviesSlice');
jest.mock('../../redux/moviesSelectors', () => ({
  selectListMovies: () => []
}));

const getComponent = (store, props) => (
  <MemoryRouter>
    <Route>
      <reactRedux.Provider store={store}>
        <ListMovies {...props} />
      </reactRedux.Provider>
    </Route>
  </MemoryRouter>
);

describe('ListMovies component', () => {
  test('render ListMovies snapshot', () => {
    const store = configureStore()();
    const component = getComponent(store);
    const { asFragment } = render(component);

    expect(asFragment(component)).toMatchSnapshot();
  });

  test('should call getListMovies on change prop', () => {
    const store = configureStore()();
    const props = {
      search: 'search', genre: 'genre', sortBy: 'sortBy'
    };
    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(jest.fn());

    const { rerender } = render(getComponent(store));
    rerender(getComponent(store, props));

    expect(getListMovies).toHaveBeenCalledWith({
      search: props.search,
      sortBy: props.sortBy,
      sortOrder: 'desc',
      searchBy: 'title',
      filter: props.genre,
    });

    rerender(getComponent(store, { ...props, genre: 'all' }));

    expect(getListMovies).toHaveBeenCalledWith({
      search: props.search,
      sortBy: props.sortBy,
      sortOrder: 'desc',
      searchBy: 'title',
      filter: '',
    });

  });

  // test('should call setSelectedOption', () => {
  //   const movie = { id: 'id1', genres: [] };
  //   const store = configureStore()();
  //   const mockSetSelectedOption = jest.fn();

  //   jest.spyOn(reactRedux, 'useSelector').mockReturnValue([movie]);
  //   jest.spyOn(React, 'useState').mockReturnValue(([{}, mockSetSelectedOption]));
  //   const { getByTestId } = render(getComponent(store));

  //   userEvent.click(getByTestId('selectOption'), 'EDIT');

  //   expect(mockSetSelectedOption).toHaveBeenCalledWith({ movie, type: 'EDIT' });

  // });
});