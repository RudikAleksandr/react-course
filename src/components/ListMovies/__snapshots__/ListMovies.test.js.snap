// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListMovies component render ListMovies snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex-column mb-4 listMoviesContainer row"
  >
    <div
      class="pl-0 mt-4 mb-4 col"
    >
      <span
        class="countMovies"
      >
        0
      </span>
      <span
        class="moviesFoundLabel"
      >
        movies found
      </span>
    </div>
    <div
      class="col"
    >
      <div
        class="noMovieFound"
      >
        <h2
          class="noMovieFoundLabel"
        >
          No Movie Found
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
